client/my-sites/term-tree-selector/terms.jsx

```diff
@@ -0,0 +1,413 @@
+/**
+ * External dependencies
+ */
+import React, { PropTypes } from 'react';
+import { connect } from 'react-redux';
+import classNames from 'classnames';
+import unescapeString from 'lodash/unescape';
+import includes from 'lodash/includes';
+import { localize } from 'i18n-calypso';
+import debounce from 'lodash/debounce';
+import range from 'lodash/range';
+import VirtualScroll from 'react-virtualized/VirtualScroll';
+import difference from 'lodash/difference';
+
+/**
+ * Internal dependencies
+ */
+import analytics from 'lib/analytics';
+import NoResults from './no-results';
+import Search from './search';
+import QueryTerms from 'components/data/query-terms';
+import { getSelectedSiteId } from 'state/ui/selectors';
+import {
+	isRequestingTermsForQueryIgnoringPage,
+	getTermsLastPageForQuery,
+	getTermsForQueryIgnoringPage,
+	getTermsHierarchyForQueryIgnoringPage
+} from 'state/terms/selectors';
+
+/**
+ * Constants
+ */
+const SEARCH_DEBOUNCE_TIME_MS = 500;
+const DEFAULT_TERMS_PER_PAGE = 100;
+const LOAD_OFFSET = 10;
+const ITEM_HEIGHT = 25;
+
+const TermTreeSelectorList = React.createClass( {
+
+	propTypes: {
+		terms: PropTypes.array,
+		termsHierarchy: PropTypes.array,
+		taxonomy: PropTypes.string,
+		multiple: PropTypes.bool,
+		selected: PropTypes.array,
+		search: PropTypes.string,
+		siteId: PropTypes.number,
+		translate: PropTypes.func,
+		defaultTermId: PropTypes.number,
+		lastPage: PropTypes.number,
+		onSearch: PropTypes.func,
+		onChange: PropTypes.func
+	},
+
+	getInitialState() {
+		return {
+			searchTerm: '',
+			requestedPages: [ 1 ]
+		};
+	},
+
+	getDefaultProps() {
+		return {
+			analyticsPrefix: 'Category Selector',
+			searchThreshold: 8,
+			loading: true,
+			terms: [],
+			onSearch: () => {},
+			onChange: () => {},
+			onNextPage: () => {}
+		};
+	},
+
+	componentWillMount() {
+		this.itemHeights = {};
+		this.hasPerformedSearch = false;
+		this.virtualScroll = null;
+
+		this.queueRecomputeRowHeights = debounce( this.recomputeRowHeights );
+		this.debouncedSearch = debounce( () => {
+			this.props.onSearch( this.state.searchTerm );
+		}, SEARCH_DEBOUNCE_TIME_MS );
+	},
+
+	componentWillReceiveProps( nextProps ) {
+		if ( nextProps.taxonomy !== this.props.taxonomy ) {
+			this.setState( {
```

Comment:
Could get away with `this.setState( this.getInitialState() )` here but sometimes can lead to maintenance issues if someone revises `getInitialState` without understanding the repercussions.